smolyay/basis/basis.py

```diff
@@ -0,0 +1,249 @@
+        if len(self._points) == 0:
+            self._compute_points()
+        return self._points
+
+    @property
+    def n(self):
+        """degree of polynomial"""
+        return self._n
+
+    @n.setter
+    def n(self,n):
+        if self._n != n:
+            self._n = n
+            self._points = []
+
+    def _compute_points(self):
+        """Compute extrema of Chebyshev polynomial of the first kind"""
+        if self._n == 0:
+            self._points = [0]
+        else:
+            for i in range(0,self._n+1):
+                temp = round(-math.cos(math.pi*(i)/(self._n)),15) + 0
+                self._points.append(temp)
+
+    def __call__(self,x):
+        """Terms of basis function
+        Returns the output of an nth degree Chebyshev polynomial of the
+        first kind with input x
+
+        Parameters
+        ----------
+        x : float
+            input
+
+        Returns
+        -------
+        y : float
+            output
+        """
+        if self._n == 0:
+            return 1
+        elif self._n == 1:
+            return x
+        else:
+            k_lim = self._n//2
+            answer = 0
+            for k in range(0,k_lim+1):
+                answer += math.comb(self._n,2*k)*((x**2 - 1)**k)*(x**(self._n-2*k))
+            return answer
+
+
+class BasisFunctionSet(abc.ABC):
+    """Set of basis functions
+    Creates a framework for a set of basis functions to be used as the
+    building blocks for a surrogate model.
+    ``basis_set`` is a list of BasisFunction objects
+    ``sample_flag`` is a list of boolean equal in length to basis_set and
+    an index in the list is true if the object in basis_set at the same
+    index is one that points should be sampled from
+    ``basis_function`` is the BasisFunction child class that populates
+    basis_set
+    ''all_points'' 1D points taken from the BasisFunction objects specified
+    by sample_flag
+    ``need_update`` is a boolean parameter that specifies when properties
+    of the class need to be recomputed
+
+    """
+
+    def __init__(self,sample_flag,basis_function):
+        self._sample_flag = sample_flag
+        self._basis_function = basis_function
+        self._basis_set = []
+        self._all_points = []
+        self._need_update = True
+
+    @property
+    def all_points(self):
+        """All points for basis function set at some level of precision"""
+        if self._need_update:
+            self._update()
+        return self._all_points
+
+    @property
+    def basis_set(self):
+        """list of BasisFunction objects"""
+        if self._need_update:
+            self._update()
+        return self._basis_set
+
+    @property
+    def sample_flag(self):
+        """list of what objects in basis_set to sample points from"""
+        return self._sample_flag
+
+    @sample_flag.setter
+    def sample_flag(self,sample_flag):
+        self._sample_flag = sample_flag
+        self._need_update = True
```

Comment:
I don't quite follow the purpose/usage of this, could you try explain more?

Reply:
The points that BasisFunctionSet is supposed to get are taken from specific basis functions. In the case of the Chebyshev polynomial, it takes the extrema of polynomials of degree 0, 2, 4, 8, 16, and so on. `basis_set` stores a list of objects from a chebyshev polynomial of degree 0 to the max polynomial degree its supposed to sample extrema from. 

However, BasisFunctionSet has no way of knowing which objects in` basis_set` need to be sampled, so there needs to be a variable that tells the class that information. 

`sample_flag = [True, False, True, False, True, False, False, False, True]' means sample points from the objects in 'basis_set' at indexes 0, 2, 4, and 8. Because of this setup, 'basis_set' will always be the same length as 'sample_flag'.

I've been talking to Mohammadreza and think that it would be a good idea to replace sample_flag with a list of the indexes to sample points from. Something like `sample_index = [0,2,4,8]` We figured we should first hold it off until discussing it in our next meeting.

Reply:
Ah OK, I see. I definitely like the second solution, it is easier to understand. I agree we should talk about how the `BasisFunctionSet` is going to be used subsequently before deciding on a specific interface. There may be some other options in addition to these two.